### PR TITLE
DOC-9218: [How-to Guides for Server Developers] Cost-Based Optimizer

### PR DESCRIPTION
**Pull request by simon-dew**
The following draft documentation is ready for review:

[Developer Guides › Optimizing Performance](https://simon-dew.github.io/docs-site/DOC-9218/server/current/guides/optimize.html)
[Developer Guides › Optimizing Performance › Cost-Based Optimizer](https://simon-dew.github.io/docs-site/DOC-9218/server/current/guides/cbo.html)

Docs issue: [DOC-9218](https://issues.couchbase.com/browse/DOC-9218)

### CODE REVIEW

File: modules/n1ql/pages/n1ql-language-reference/statistics-delete.adoc [hunk 1]
```diff
@@ -32,14 +32,12 @@ Deleting all statistics for a keyspace turns off the cost-based optimizer for al
...
-[subs="normal"]
+[source,ebnf]
 ----
-update-statistics-delete ::=
-  ( UPDATE STATISTICS [ FOR ] | ANALYZE [ KEYSPACE | COLLECTION ] )
-    <<keyspace-ref>> <<delete-clause>>
+include::partial$grammar/utility.ebnf[tag=update-statistics-delete]
 ----
 
-image::n1ql-language-reference/update-statistics-delete.png["( 'UPDATE' 'STATISTICS' 'FOR'? | 'ANALYZE' ( 'KEYSPACE' | 'COLLECTION')? ) keyspace-ref delete-clause"]
```
Comment (osfameron) [inline modules/n1ql/pages/n1ql-language-reference/statistics-delete.adoc:42]: Previous contents were ALT text?
(I don't think we're very consistent with those, probably another one for a style-guide discussion sometime...)

File: modules/guides/pages/indexes.adoc [hunk 2]
```diff
@@ -50,10 +50,4 @@ include::select-index.adoc[tags=intro]
...
 
-* xref:drop-index.adoc[Dropping Indexes]
-
-== Index Advisor
-
-include::index-advisor.adoc[tags=intro]
-
-* xref:index-advisor.adoc[Index Advisor]
\ No newline at end of file
+* xref:drop-index.adoc[Dropping Indexes]
```
Comment (osfameron) [inline modules/guides/pages/indexes.adoc]: As the Index Advisor is moving *out* of this section, isn't keeping the *link* here even more important now, as people might still be looking here for this content?

File: modules/guides/pages/optimize.adoc [hunk 3]
```diff
@@ -0,0 +1,41 @@
+= Optimizing Performance
+:page-role: tiles -toc
+:page-pagination: next
+:description: This guide explains how to optimize the performance of N1QL queries using prepared statements, the index advisor, and the cost-based optimizer.
+:!sectids:
+
+// Pass through HTML styles for this page.
+
+ifdef::basebackend-html[]
```
Comment (osfameron) [inline modules/guides/pages/optimize.adoc:9]: Ah, so this is the ugly hack for CSS styles?
It's good to know (or not, arguably) but happy to pair/review extracting this out to docs-ui as we did for the REST API stuff.

**Review by osfameron — APPROVED** [review]
As mentioned in Slack, giving a 👍 to unblock if required, but with caveat that I don't understand the content / some of the markup complexities well enough for this to be very well informed. 
I'm available to do a content/code walkthrough on Zoom if that's more helpful.

File: modules/guides/pages/indexes.adoc (hunk 2 above)
Comment (simon-dew) [reply modules/guides/pages/indexes.adoc]: Excellent point

File: modules/guides/pages/cbo.adoc [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+xref:n1ql-language-reference/merge.adoc[MERGE], and
+xref:n1ql-language-reference/insert.adoc[INSERT INTO with SELECT] queries.
+
+include::partial$before-you-begin.adoc[tag=body]
+include::partial$query-tools.adoc[tags=body;!thumbs]
+
+== Activating the Cost-Based Optimizer
+
+The Cost-Based Optimizer is active by default.
+This section covers shows how to activate or deactivate the Cost-Based Optimizer for a request.
```
Comment (maria-robobug) [inline modules/guides/pages/cbo.adoc]: Probably just a typo, `covers shows`

Suggestion:
```suggestion
This section shows how to activate or deactivate the Cost-Based Optimizer for a request.
```

File: modules/guides/pages/cbo.adoc [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+* xref:c-sdk:howtos:n1ql-queries-with-sdk.adoc[C]
+| xref:dotnet-sdk:howtos:n1ql-queries-with-sdk.adoc[.NET]
+| xref:go-sdk:howtos:n1ql-queries-with-sdk.adoc[Go]
+| xref:java-sdk:howtos:n1ql-queries-with-sdk.adoc[Java]
+| xref:nodejs-sdk:howtos:n1ql-queries-with-sdk.adoc[Node.js]
+| xref:php-sdk:howtos:n1ql-queries-with-sdk.adoc[PHP]
+| xref:python-sdk:howtos:n1ql-queries-with-sdk.adoc[Python]
+| xref:ruby-sdk:howtos:n1ql-queries-with-sdk.adoc[Ruby]
+| xref:scala-sdk:howtos:n1ql-queries-with-sdk.adoc[Scala]
```
Comment (maria-robobug) [inline modules/guides/pages/cbo.adoc:209]: Should probably add the Kotlin SDK to these references as well.

```
| xref:kotlin-sdk:howtos:n1ql-queries.adoc[Kotlin]
```

File: modules/guides/pages/optimize.adoc (hunk 3 above)
Comment (simon-dew) [reply modules/guides/pages/optimize.adoc:9]: Yes. I think this one has been tested enough that I should probably put it into docs-ui 😄

File: modules/n1ql/pages/n1ql-language-reference/statistics-delete.adoc (hunk 1 above)
Comment (simon-dew) [reply modules/n1ql/pages/n1ql-language-reference/statistics-delete.adoc:42]: The content of the EBNF code block above _is_ effectively the alt text for these syntax diagrams. I should probably add alt text to the images saying so.

File: modules/guides/pages/cbo.adoc [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+{description}
+
+== Introduction
+
+// tag::intro[]
+In Couchbase Server Enterprise Edition, the Cost-Based Optimizer enables the Query service to create the most efficient plan to execute a query.
+// end::intro[]
+
+The Cost-Based Optimizer can generate a query plan for
+xref:n1ql-language-reference/selectintro.adoc[SELECT],
```
Comment (maria-robobug) [inline modules/guides/pages/cbo.adoc]: I tested these out in the page preview and the refs/links don't work. 
From what I can see [here](https://github.com/couchbase/docs-server/tree/release/7.1/modules/n1ql/pages/n1ql-language-reference), this should be:

```suggestion
xref:n1ql:n1ql-language-reference/selectintro.adoc[SELECT],
```
Same for the others below.

**Review by maria-robobug — APPROVED** [review]
👍 LGMT, just some minor comments.

File: modules/guides/pages/cbo.adoc (hunk 6 above)
Comment (simon-dew) [reply modules/guides/pages/cbo.adoc]: Done

File: modules/guides/pages/cbo.adoc (hunk 4 above)
Comment (simon-dew) [reply modules/guides/pages/cbo.adoc]: Done

File: modules/guides/pages/cbo.adoc (hunk 5 above)
Comment (simon-dew) [reply modules/guides/pages/cbo.adoc:209]: Good point, but probably outside the scope of this PR. I'd need to know what to do about areas where there is no Kotlin SDK documentation at the moment, e.g. subdoc operations, indexes.

File: modules/n1ql/pages/n1ql-language-reference/statistics-delete.adoc (hunk 1 above)
Comment (simon-dew) [reply modules/n1ql/pages/n1ql-language-reference/statistics-delete.adoc:42]: I'll do this outside the scope of this PR.

File: modules/guides/pages/indexes.adoc (hunk 2 above)
Comment (simon-dew) [reply modules/guides/pages/indexes.adoc]: Done

**Pull request merged**